megamek/src/megamek/client/bot/princess/Princess.java

```diff
@@ -1464,6 +1485,102 @@ protected void initFiring() {
             methodEnd(getClass(), METHOD_NAME);
         }
     }
+    
+    /**
+     * Function with side effects. Retrieves the move path collection we want
+     * the entity to consider. Sometimes it's the standard "circle", sometimes it's pruned long-range movement paths
+     */
+    public List<MovePath> getMovePathsAndSetNecessaryTargets(Entity mover, boolean forceMoveToContact) {
+        // if the mover can't move, then there's nothing for us to do here, let's cut out.
+        if(mover.isImmobile()) {
+            return new ArrayList<MovePath>();
+        }
+        
+        BehaviorType behavior = forceMoveToContact ? BehaviorType.MoveToContact : unitBehaviorTracker.getBehaviorType(mover, this);
+        boardClusterTracker.clearMovableAreas();
+        boardClusterTracker.updateMovableAreas(mover);
+        
+        // basic idea: 
+        // if we're "in battle", just use the standard set of move paths
+        // if we're trying to get somewhere
+        //  - sort all long range paths by "mp cost" (actual MP + how long it'll take to do terrain leveling)
+        //  - set the first terrain/building as 'strategic target' if the shortest path requires terrain leveling
+        //  - if the first strategic target is in LOS at the pruned end of the shortest path,  
+        //      then we actually return the paths for "engaged" behavior
+        // TODO: special case is 'move to contact' mode, which should evaluate the long range paths if none of the
+        // basic paths produce potential damage 
+        
+        switch(behavior) {
+        case Engaged:
+            return getPrecognition().getPathEnumerator()
+                    .getUnitPaths()
+                    .get(mover.getId());
+        case MoveToDestination:
+        case MoveToContact:
+        case ForcedWithdrawal:
+        default:
+            List<BulldozerMovePath> bulldozerPaths = getPrecognition().getPathEnumerator().
+                getLongRangePaths().get(mover.getId());
+            
+            // for whatever reason (most likely it's wheeled), there are no long-range paths for this unit, 
+            // so just have it mill around in place as usual. Also set the behavior to "engaged"
+            // so it doesn't hump the walls due to "self preservation mods"
+            if ((bulldozerPaths == null) || (bulldozerPaths.size() == 0)) {
+                getUnitBehaviorTracker().overrideBehaviorType(mover, BehaviorType.NoPathToDestination);
+                return getPrecognition().getPathEnumerator()
+                        .getUnitPaths()
+                        .get(mover.getId());
+            }
+            
+            Collections.sort(bulldozerPaths, new BulldozerMovePath.MPCostComparator());
+            
+            // if the quickest route needs some terrain adjustments, let's get working on that
+            Targetable levelingTarget = null;
+            
+            if (bulldozerPaths.get(0).needsLeveling()) {
```

Comment:
Will we always have paths here?

Reply:
Yes, on line 1528 there's a "null or size 0" check.